### PR TITLE
fix the bug,  can not update topic when  the update  topicName is contained by an existed topic as a part

### LINKED ISSUE #11685: can not update topic when  the update  topicName is contained by an existed topic as a part
**Describe the bug**
can not update topic when  the update  topicName is contained by an existed topic as a part

image

### PR DESCRIPTION
**Pull request by Nicklee007**
Fixes #11685
 
validatePartitionTopicUpdate use contain to check if there has a exist topic will cause conflict, which  will cause a failed when exist a topic which contain the new topic's prefix and we want to update the new topic partition;

 we have a those topic 
"persistent://public/default/113p-partition-0"
"persistent://public/default/113p-partition-1"
"persistent://public/default/113p-partition-2"
"persistent://public/default/3p-partition-0"

when we want to update topic 3p to more partitions ,we failed because "persistent://public/default/113p-partition-0" contain "3p-partition"

### Modifications

use the startwith to check if exist the same topic.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -3773,9 +3773,9 @@ private void validatePartitionTopicUpdate(String topicName, int numberOfPartitio
         TopicName partitionTopicName = TopicName.get(domain(), namespaceName, topicName);
         PartitionedTopicMetadata metadata = getPartitionedTopicMetadata(partitionTopicName, false, false);
         int oldPartition = metadata.partitions;
-        String prefix = topicName + TopicName.PARTITIONED_TOPIC_SUFFIX;
+        String prefix = partitionTopicName.getPartitionedTopicName() + TopicName.PARTITIONED_TOPIC_SUFFIX;
         for (String exsitingTopicName : existingTopicList) {
-            if (exsitingTopicName.contains(prefix)) {
+            if (exsitingTopicName.startsWith(prefix)) {
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: This change won't work properly. The topic name will be in format `persistent://tenant/cluster/namespace/topicName` .

Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: There is a PR #11148 in progress in this same area. I'd suggest contributing to #11148 to fix this issue #11685. One form of contributing is making sure that this problem gets fixed with the refactored implementation for checking topic existence.

**Review by lhotari — CHANGES_REQUESTED** [review]

**Comment by Anonymitaet** [comment]
Thanks for your contribution. For this PR, do we need to update docs?

(The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc-related info in this and future PR descriptions? Thanks)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (Nicklee007) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: > This change won't work properly. The topic name will be in format `persistent://tenant/cluster/namespace/topicName` .

yes I have considered that.  in this PR，I changed to the prefix = topicName + TopicName.PARTITIONED_TOPIC_SUFFIX, it's the full topicName contain 'persistent://tenant/cluster/namespace/topicName';

Comment (gaozhangmin) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: > There is a PR #11148 in progress in this same area. I'd suggest contributing to #11148 to fix this issue #11685. One form of contributing is making sure that this problem gets fixed with the refactored implementation for checking topic existence.

it's full name `String prefix = partitionTopicName.getPartitionedTopicName() + TopicName.PARTITIONED_TOPIC_SUFFIX`

Comment (Nicklee007) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: > This change won't work properly. The topic name will be in format `persistent://tenant/cluster/namespace/topicName` .

changed to the prefix = partitionTopicName.getPartitionedTopicName() + TopicName.PARTITIONED_TOPIC_SUFFIX , it's the full topicName contain 'persistent://tenant/cluster/namespace/topicName';

**Comment by Nicklee007** [comment]
> Thanks for your contribution. For this PR, do we need to update docs?
> 
> (The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc-related info in this and future PR descriptions? Thanks)

 It's fix a bug and don't change any system function,  I think not need to update docs.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (gaozhangmin) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: > same are

I don't think it's the same area, `validatePartitionTopicUpdate` only used when updating partition of topic. it's not the same logic.

**Review by gaozhangmin — CHANGES_REQUESTED** [review]
miss space between topic and exsitingTopicName

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3778]: @lhotari I don't think this is the same issue as #11148 wants to fix. This is a bug here, #11148 is trying to improve the topic creation logic.

**Review by codelipenghui — COMMENTED** [review]
The change looks good to me, @Nicklee007 Could you please add some tests for the new change?

**Comment by Nicklee007** [comment]
> The change looks good to me, @Nicklee007 Could you please add some tests for the new change?



> new

thanks, I have  added the test for update the partitioned topic which a part is contained in old topicName.

**Comment by Nicklee007** [comment]
> The change looks good to me, @Nicklee007 Could you please add some tests for the new change?

@codelipenghui thanks, I have added the test for update the partitioned topic which a part topicName is contained in another old topicName.

**Review by codelipenghui — APPROVED** [review]

**Review by hangc0276 — APPROVED** [review]

**Comment by hangc0276** [comment]
@lhotari  Would you please review this PR again? I am cutting 2.8.1, i'd like to include this pr in 2.8.1, thanks.

**Review by lhotari — APPROVED** [review]
Good work @Nicklee007

**Pull request merged**